src/Bounds.cpp

```diff
@@ -2103,19 +2103,38 @@ class BoxesTouched : public IRGraphVisitor {
         }
     };
 
+    // Build an ordered list of the unique children of 'name', such that every
```

Comment:
This looks like it still does a full depth-first traversal graph, visiting nodes multiple times even though we know their children will never be pushed onto 'ordered'. I think if (result.second) is false you can just return, with no need to visit the children again. Or how about the following non-recursive alternative. I think the only different is that the children get visited in reverse order, but the children are still visited before the parents.

```

    void trim_scope_push(const string &name, const Interval &bound, vector<LetBound> &let_bounds) {
        scope.push(name, bound);

        vector<string> pending;
        set<string> unique;
        pending.push_back(name);
        unique.insert(name);
        while (!pending.empty()) {
            string next = std::move(pending.back());
            pending.pop_back();
            for (const auto &v : children[get_var_instance(next)]) {
                if (unique.insert(v.var).second) {
                    pending.push_back(v.var);
                    string max_name = unique_name('t');
                    string min_name = unique_name('t');

                    let_bounds.insert(let_bounds.begin(), LetBound(v.var, min_name, max_name));

                    Type t = let_stmts.get(v.var).type();
                    Interval b = Interval(Variable::make(t, min_name), Variable::make(t, max_name));
                    scope.push(v.var, b);
                }
            }
        }
    }

```

Reply:
Yeah, it visits everything, it just uses a set to filter out the dupes, so it isn't as efficient as it could be. I'll give your suggestion a try.

Reply:
One possible gotcha with yours: the toplevel 'name' passed in needs to be pushed onto scope, but I don't think it should be added to let_bounds, as while let_stmts are expected for all the children, I don't think it's expected for the name itself.

Reply:
Edit: ah, nevermind, I see this explicitly only processes the children. This looks better than mine, I'll update the PR.